app/scripts/modules/core/src/presentation/modal/Modal.tsx

```diff
@@ -0,0 +1,48 @@
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import { CSSTransition } from 'react-transition-group';
+
+const { useMemo } = React;
+
+import { useEventListener, useContainerClassNames, useLatestCallback } from 'core/presentation';
+
+import { ModalContext } from './ModalContext';
+import * as styles from './Modal.module.css';
+
+export interface IModalProps {
+  isOpen: boolean;
+  onRequestClose?: () => any;
+  children?: React.ReactNode;
+}
+
+export const Modal = ({ onRequestClose, isOpen, children }: IModalProps) => {
+  useContainerClassNames(isOpen ? [styles.backdropBlurEffect] : []);
+
+  const keydownCallback = ({ keyCode }: KeyboardEvent) => {
+    if (keyCode === 27) {
+      onRequestClose();
+    }
+  };
+  useEventListener(document, 'keydown', isOpen ? keydownCallback : null);
+
+  const memoizedRequestClose = useLatestCallback(onRequestClose);
+  const modalContext = useMemo(() => ({ onRequestClose: memoizedRequestClose }), []);
+
+  return (
+    <ModalContext.Provider value={modalContext}>
```

Comment:
Question.  Thought.  Whatever.  

I'm generally hesitant to add context because it can obfuscate the flow of data.
How would you feel about using a render prop instead?

```js
<Modal render={({ onRequestClose }) => (
  <ModalHeader onClose={onRequestClose} />
)} />

Reply:
Funny enough, I went in circles a couple times re: whether to make the plumbing of the header close behavior something you glue together yourself vs. it being done for you under the hood.
After spending time with these components and thinking about other Modal components, how Formik uses render props vs. context, etc., I came around to thinking this behavior specifically is a good candidate for context. Specifically:

- Much like the wiring up of Formik's `Field` components to Formik state, it's really just "dumb" plumbing/noise that adds no value to the mental model of a consumer — it's going to work exactly the same way, every single time, and the most you as an app developer can really do to modify it is to mis-use it in some broken or sub-optimal way
- Letting the component system do the plumbing itself enforces a bit of separation between app code and the way modals use app code to present things on the screen/respond to inputs. I fully expect we will be layering higher level kinds of modals on top of this base modal component system, and it's entirely possible we'll want to do the same thing we're doing here with `ModalHeader` in other sub-components (dismiss buttons, links or other interactive chunks inside the body, etc.)
- One of the things I've observed about our use of `react-bootstrap` modals is that the interaction mechanics can be inconsistent in ways which are frustrating as an end user (and a little confusing as an engineer). Depending on how people choose to wire up the components and event handlers, sometimes you get `esc` key handling and sometimes not, sometimes you get a close button that does things a different way from the `esc` code path, we even have some modals that don't add close buttons at all (and I think there are some that are hand-built rather than using the `ModalClose` component). I'd love for the component/design systems we build to take the burden and cognitive load of maintaining consistent UX choices away from individuals and codify them into something that we can all confidently use the same way without needing to make the repetitive choices others have made hundreds of times already

After this little journey of self discovery, I think my personal opinion has become that if something:
  1) is intended and expected to be boilerplate plumbing that works the same way every time
  2) doesn't have a major impact on the mental model of the person consuming/reading it
  3) would benefit from being consistently applied
  4) isn't complicated enough in its data flow/state that hiding it is confusing 

then context is a good call. If any of the above aren't true or have sufficiently large asterisks, it's worth letting folks plumb them together manually using something like a render prop. I generally think that matches the approach Formik takes, but also always happy to eat my words if this proves to be annoying / not useful later on. I sincerely think this is a choice that has a low cost to change/walk back, because the refactor of consuming code would be pretty straightforward as long as the Modal components handle memoizing callbacks themselves.